Add EIP-6493: SSZ Transaction Signature Scheme (#6493)
ethereum/EIPs: EIPS/eip-6493.md

---
eip: 6493
title: SSZ Transaction Signature Scheme
description: Signature scheme for SSZ transactions
author: Etan Kissling (@etan-status), Matt Garnett (@lightclient), Vitalik Buterin (@vbuterin)
discussions-to: https://ethereum-magicians.org/t/eip-6493-ssz-transaction-signature-scheme/13050
status: Draft
type: Standards Track
category: Core
created: 2023-02-24
requires: 155, 2124, 2718
---

## Abstract

This EIP defines a signature scheme for [Simple Serialize (SSZ)](https://github.com/ethereum/consensus-specs/blob/67c2f9ee9eb562f7cc02b2ff90d92c56137944e1/ssz/simple-serialize.md) encoded transactions.

## Motivation

> Existing [EIP-2718](./eip-2718.md) transaction types are encoded in the RLP format, and then hashed using keccak256 for signing. For new transaction types that are encoded in the SSZ format, it is ideomatic to base their signature hash and their public identifier on `hash_tree_root` instead.

Reviewer: ```suggestion
Existing [EIP-2718](./eip-2718.md) transaction types are encoded in the RLP format, and then hashed using keccak256 for signing and to generate a unique transaction reference/public identifier. 
For new transaction types that are encoded in the SSZ format, it is ideomatic to base their signature hash and their public identifier on `hash_tree_root` instead.
```

## Specification

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.

### SSZ transaction schema

For each SSZ transaction type, its specification defines an SSZ type that contains all transaction data to be signed.

```python
class XyzTransaction(View):
    ...
```

> Each specification also defines an SSZ type that represents the transaction's signature.

Reviewer: ```suggestion
The specification also defines an SSZ type that represents the transaction's signature.
```

```python
class XyzSignature(View):
    ...
```

The signed transaction combines the unsigned transaction data with its signature.

```python
class XyzSignedTransaction(Container):
    message: XyzTransaction
    signature: XyzSignature
```

The unique transaction identifier is defined as this signed container's SSZ `hash_tree_root`.

> When representing the transaction as part of the [execution block header's `txs-root`](https://github.com/ethereum/devp2p/blob/bd17dac4228c69b6379644355f373669f74952cd/caps/eth.md#block-encoding-and-validity) Merkle-Patricia Trie, that unique transaction identifier is embedded instead of the transaction's raw network representation.

Reviewer: ```suggestion
When representing such a  transaction as part of the [execution block header's `txs-root`](https://github.com/ethereum/devp2p/blob/bd17dac4228c69b6379644355f373669f74952cd/caps/eth.md#block-encoding-and-validity) Merkle-Patricia Trie, this unique transaction identifier is embedded instead of the transaction's raw network representation.
```

Reviewer 2: Typo fix:
```
When representing such a transaction as part of the [execution block header's `txs-root`](https://github.com/ethereum/devp2p/blob/bd17dac4228c69b6379644355f373669f74952cd/caps/eth.md#block-encoding-and-validity) Merkle-Patricia Trie, this unique transaction identifier is embedded instead of the transaction's raw network representation.
```

    return compute_execution_domain(domain_type, tx_type_fork_version, genesis_hash, chain_id)
```

The [`Root`](https://github.com/ethereum/consensus-specs/blob/67c2f9ee9eb562f7cc02b2ff90d92c56137944e1/specs/phase0/beacon-chain.md#custom-types) to sign is computed using [`compute_signing_root`](https://github.com/ethereum/consensus-specs/blob/67c2f9ee9eb562f7cc02b2ff90d92c56137944e1/specs/phase0/beacon-chain.md#compute_signing_root) based on the unsigned transaction's `hash_tree_root` and the additional information about the transaction type.

## Rationale

### Why not keccak256?

SSZ and RLP objects encode differently. Namely, in an encoded SSZ transaction, it is not guaranteed that the `chain_id` and the signature are at the same location as in the RLP transaction. This could be problematic if two different networks accidentally use the same [EIP-2718](./eip-2718.md) transaction type number to define an RLP encoded transaction type on one network, but an SSZ encoded transaction type on the other. A signed transaction on one network could suddenly become a valid transaction on the other network.

### Why the specific domain type value?

> [`DomainType`](https://github.com/ethereum/consensus-specs/blob/67c2f9ee9eb562f7cc02b2ff90d92c56137944e1/specs/phase0/beacon-chain.md#custom-types) is used in consensus to isolate signing domains for validator BLS signatures. So far, execution uses secp256k1 ECDSA signatures instead, so it is not strictly necessary to isolate consensus and execution domains from each other. However, with 4 bytes, avoiding collisions across layers is trivially possible and might be useful in future use cases.

Reviewer: ```suggestion
[`DomainType`](https://github.com/ethereum/consensus-specs/blob/67c2f9ee9eb562f7cc02b2ff90d92c56137944e1/specs/phase0/beacon-chain.md#custom-types) is used in consensus to isolate signing domains for validating BLS signatures. So far, execution uses secp256k1 ECDSA signatures instead, so it is not strictly necessary to isolate consensus and execution domains from each other. However, with 4 bytes, avoiding collisions across layers is trivially possible and might be useful in future use cases.
```

CHAIN_ID = uint256(424242)

# Example SSZ transaction
EXAMPLE_TX_TYPE = TransactionType(0xab)
EXAMPLE_TX_TYPE_FORK_VERSION = Version('0x12345678')

class ExampleTransaction(Container):
    chain_id: uint256
    nonce: uint64
    max_fee_per_gas: uint256
    gas: uint64
    tx_to: ExecutionAddress
    tx_value: uint256

class ExampleSignature(ByteVector[65]):
    pass

class ExampleSignedTransaction(Container):
    message: ExampleTransaction
    signature: ExampleSignature

def compute_example_sig_hash(message: ExampleTransaction) -> Hash32:
    domain = compute_transaction_domain(
        EXAMPLE_TX_TYPE,
        EXAMPLE_TX_TYPE_FORK_VERSION,
        GENESIS_HASH,
        CHAIN_ID,
    )
    return compute_signing_root(message, domain)

def compute_example_tx_hash(signed_tx: ExampleSignedTransaction) -> Hash32:
    return signed_tx.hash_tree_root()

# Example transaction
message = ExampleTransaction(
    chain_id=CHAIN_ID,
    nonce=42,
    max_fee_per_gas=69123456789,
    gas=21000,
    tx_to=ExecutionAddress(bytes.fromhex('d8da6bf26964af9d7eed9e03e53415d37aa96045')),
    tx_value=3_141_592_653,
)

> sig_hash = compute_example_sig_hash(message)

Reviewer: ```suggestion
sig_root = compute_example_sig_root(message)
```
would this be clearer to convey that this goes into signature calc
